node/flatpak_node_generator/providers/npm.py

```diff
@@ -64,8 +64,10 @@ def process_dependencies(
             elif version.startswith('file:'):
                 source = LocalSource(path=version[len('file:') :])
             else:
-                integrity = Integrity.parse(info['integrity'])
-                source = ResolvedSource(resolved=info['resolved'], integrity=integrity)
+                source = RegistrySource(
+                    resolved=info.get('resolved'),
```

Comment:
ack, this is making me want to bring back `UnresolvedRegistrySource` which was like, [just removed](https://github.com/flatpak/flatpak-builder-tools/pull/289/commits/7fc0cf2ecf7f20974d79d08a3d51fab1954bd826) :sweat_smile: 

Since right now, ResolvedSource/RegistrySource is used to hold two different things:

1. an unresolved package, with only an integrity to go off of
2. a fully resolved package

In npm-land, a source transitions from 1->2 with `resolve_source`, after which we know it's resolved. In yarn-land, it's *always* resolved...but in both cases, now we need to handle the case where `resolved is None`.

So let's hypothetically say `UnresolvedRegistrySource` is brought back, but now with a single member `integrity: Integrity`. and `resolve_source` takes a Package with an `UnresolvedRegistrySource` inside (just like before). Then, *all* of the assertions on `resolved` could actually be removed, because it's virtually guaranteed our source values are always resolved for us already.

(Of course really, the assertions etc are arguably off-topic for this PR anyway, but at minimum I feel like bringing back `UnresolvedRegistrySource` for this purpose is a bit nicer anyway!)

(also ngl I feel like the original intent of `UnresolvedRegistrySource` in the first place was something like this? but it's been so long I forgot the details...)

Reply:
Maybe make make `*Source` classes datacalsses instead of NamedTuples, then? If bring `UnresolvedRegistrySource` back, it feels natural to use inheritance here, like
```python
class RegistrySource(PackageSource):
    pass

class UnresolvedRegistrySource(RegistrySource):
    integrity: Integrity

class ResolvedRegistrySource(RegistrySource):
    resolved: str
    integrity: Integrity
```
or even
```python
class UnresolvedRegistrySource(PackageSource):
    integrity: Integrity

class ResolvedRegistrySource(UnresolvedRegistrySource):
    resolved: str
```

Reply:
I was actually going to try sending a PR to switch these to dataclasses after this was merged, was planning to before and then I forgot. With that in mind, I'd definitely be fine with changing it here first. (the latter class layout is definitely nicer than what I had before!)

Reply:
Ok, done.
It seems natural now to switch other similar things (like `Package`) to dataclasses, but it's out of scope of this PR, so I didn't touch them.